pinot-common/src/main/java/com/linkedin/pinot/common/utils/request/RequestUtils.java

```diff
@@ -206,4 +212,43 @@ public static boolean isFitForStarTreeIndex(SegmentMetadata segmentMetadata, Fil
 
     return true;
   }
+
+  /**
+   * Must be called after a successful check of {@link #isFitForStarTreeIndex}
+   * @return true if no fasthll aggregation or rewriting successful, false otherwise
+   */
+  public static boolean performFastHllRewriting(SegmentMetadata segmentMetadata, BrokerRequest brokerRequest) {
+    List<AggregationInfo> aggregationsInfo = brokerRequest.getAggregationsInfo();
+    StarTreeMetadata starTreeMetadata = segmentMetadata.getStarTreeMetadata();
+    Map<AggregationInfo, String> fastHllAggregationInfoToDerivedColumnNameMap = new HashMap<>();
+    // Check if all fasthll derived columns exist
+    for (AggregationInfo aggregationInfo : aggregationsInfo) {
+      if (aggregationInfo.getAggregationType().toLowerCase().equals("fasthll")) {
+        String derivedColumn = extractDerivedColumn(aggregationInfo, starTreeMetadata);
```

Comment:
isFitForStarTree will give green lights to all fasthll aggregation regardless of the aggregate column, the difference in the rewriting part. When customer provides the hll column, he/she should query on it directly, Pinot does not know how to do the rewriting. But for Pinot generated column, user can query the original column, and Pinot has the knowledge of how to do the rewriting.